### DISCUSSION TITLE
Database introspection: Table metadata shows primary keys when none exist

### DISCUSSION THREAD

**Discussion by valhuber** [sqlalchemy/sqlalchemy#10329] [Usage Questions]
My project uses SQLAlchemy (2.0.15) to introspect tables and generate SQLAlchemy classes.  The class generation depends on meta data about primary keys and unique constraints.

Mostly, the tables have primary keys.  I am getting confusing results for tables without primary keys:

- `NoKey_no_unique` - no primary key or constraints:
    - metadata has a primary key with no columns
- `KeyTest_unique` - no primary key but a unique constraint:
    - meta data has no unique constraints, and 
    - primary key with no cols

Is this expected behavior?  Should I be using the `insp = inspect(engine)` api instead?

## My database

```sql
-- sqlite3 unique_keys.sqlite < unique_keys_schema.sql 


CREATE TABLE IF NOT EXISTS "KeyTest_unique"
(
    KeyName varchar(16)
);
CREATE UNIQUE INDEX index_name
    on "KeyTest_unique" (KeyName);



CREATE TABLE IF NOT EXISTS "NoKey_no_unique"
(
    just_text varchar(16)
);



CREATE TABLE IF NOT EXISTS "unique_with_key"
(
    name_unique_with_pkey varchar(16)
        constraint users_with_key_pk
            primary key
        unique
);



CREATE TABLE IF NOT EXISTS "unique_no_key"
(
    unique_no_key_name TEXT
        unique
, created integer, pass_hash test);



CREATE TABLE IF NOT EXISTS "user_notes"
(
    user_name TEXT
        constraint user_notes_users_name_fk
            references unique_no_key (unique_no_key_name),
    note      TEXT,
    id        INTEGER not null
        constraint user_notes_pk
            primary key
);

CREATE TABLE users( name TEXT, password_hash TEXT, created BIGINT, UNIQUE(name) );


CREATE TABLE IF NOT EXISTS "unique_no_key_alt" (unique_no_key_name TEXT, created integer, pass_hash test, UNIQUE(unique_no_key_name));
```

## Screen shot of table meta data

<img width="1560" alt="Screenshot 2023-09-09 at 9 18 12 PM" src="https://github.com/sqlalchemy/sqlalchemy/assets/5579631/1fac9adf-5ab9-4390-89e0-9d5be05b1fde">

**Comment by CaselIT** [comment]
Hi,

It's not very clear what are the question. 
- not all dialects can reflect all unique constraints. They may be reflected as unique index
- a table always have a primary key in sqlalchemy. If it has no columns it means that none is present in the sql

**Comment by valhuber** [comment]
I was just looking to confirm unexpected behaviors.  I expected to see UniqueConstraints, but (for sqlite) did not.

Your suggestion to look at the indexes (and test index.unique) was all I needed to solve my problem.

Thanks, and sorry if my description was not clear enough.  Many thanks for providing what I needed through the fog ;)
